sshd-core/src/main/java/org/apache/sshd/common/session/helpers/AbstractSession.java

```diff
@@ -728,7 +728,8 @@ public IoWriteFuture writePacket(Buffer buffer) throws IOException {
         }
     }
 
-    protected IoWriteFuture doWritePacket(Buffer buffer) throws IOException {
+    // NOTE: must acquire encodeLock when calling this method
```

Comment:
Why not just asserting `Thread.holdsLock(encodeLock)` at the beginning of this function?

Reply:
1.  I don't want to enforce this at the moment since this is an open-source project and thus I prefer leaving the "garbage-in-garbage-out" approach
2. The "happy flow" calls this method a lot (!!!) and I don't want to add un-necessary overhead.